src/Features/Core/Completion/Providers/AbstractSymbolCompletionProvider.cs

```diff
@@ -19,6 +19,14 @@ namespace Microsoft.CodeAnalysis.Completion.Providers
 {
     internal abstract partial class AbstractSymbolCompletionProvider : AbstractCompletionProvider
     {
+        // PERF: Many CompletionProviders dervice AbstractSymbolCompletionProvider and therefore
+        // compute identical contexts. This actually shows up on the 2-core typing test.
+        // Cache the most recent document/position/computed SyntaxContext to reduce repeat computation.
+        private static Document cachedDocument;
+        private static int cachedPosition;
+        private static AbstractSyntaxContext cachedContext;
+        private static readonly object cacheGate = new object();
```

Comment:
I would recommend doing this:
Create a new private nested class with these three fields. You can then do atomic read and update using Interlocked instructions without requiring a gate. See, for example, what I did in the VB keyword recommender recently (I'll dig up a link)
Once you've done that, it becomes (relatively) easy to extend the cache to more than one item.